### PR TITLE
Implemented gradual rollout with hotfix support

### PR DESCRIPTION
**Pull request by talkol**
Gradual rollout feature

### CODE REVIEW

File: src/dockerhub/image-poll.ts [hunk 1]
```diff
@@ -1,32 +1,123 @@
...
+    const currentVersion = this.getCurrentVersion(rolloutGroup, imageName);
+    if (!this.isUpgradeAllowed(imageVersion, currentVersion)) return;
+    this.state.applyNewImageVersion(rolloutGroup, imageName, imageVersion);
+    Logger.log(`ImagePoll: immediate update of '${imageName}:${rolloutGroup}' to ${imageVersion}.`);
+  }
+
+  performGradualRollout(rolloutGroup: string, imageName: string, imageVersion: string) {
+    // initialize first version immediately
+    const currentVersion = this.getCurrentVersion(rolloutGroup, imageName);
+    if (!currentVersion) {
```
Comment (ronnno) [inline src/dockerhub/image-poll.ts:64]: comment
// no delays before first update

File: e2e/driver.ts [hunk 2]
```diff
@@ -24,6 +24,8 @@ export class TestEnvironment {
       EthereumEndpoint: `http://ganache:7545`,
       DockerNamespace: 'orbsnetwork',
       DockerHubPollIntervalSeconds: 1,
+      RegularRolloutWindow: 2,
```
[the remark this answers is not in the document]
Comment (talkol) [reply e2e/driver.ts:27]: described in the README :|

[the remark this answers is not in the document]
Comment (talkol) [reply e2e/driver.ts:27]: I'll add

**Comment by talkol** [comment]
> `PendingVersion` and `PendingVersionTime` in the state are never read. if they are only there for troubleshooting consider removing them altogether and including in "status" page the current scheduled upgrades.
> 
> They look like something that was left behind from a previous implementation approach. unless I missed something

They are exposed via the `/status` endpoint which is reflected in the healthcheck/status.json mechanism in Boyar. Meaning we will see this info in the status page. I added it explicitly so once we release a gradual version we could easily see when every validator is planning to upgrade to it.

File: e2e/expectations-status.ts [hunk 3]
```diff
@@ -39,6 +39,20 @@ export const expectationStatus = {
         node: isValidImageVersion,
       },
     },
+    CurrentImageVersionsUpdater: {
+      main: {
+        'management-service': {
```
[the remark this answers is not in the document]
Comment (talkol) [reply e2e/expectations-status.ts:44]: prettier does it automatically, it adds the `'` because there's a dash in the middle

File: src/model/state.ts [hunk 4]
```diff
@@ -148,14 +153,33 @@ export class State {
...
-    // image version upgrades only go forward (we don't allow downgrade)
-    if (!currentVersion || Versioning.compare(imageVersion, currentVersion) > 0) {
-      this.snapshot.CurrentImageVersions[rolloutGroup][imageName] = imageVersion;
-    }
+  applyNewImageVersion(rolloutGroup: string, imageName: string, imageVersion: string) {
+    this.snapshot.CurrentImageVersions[rolloutGroup][imageName] = imageVersion;
+  }
+
+  applyNewImageVersionPollTime(time: number, rolloutGroup: string, imageName: string) {
+    const updaterStats = this.snapshot.CurrentImageVersionsUpdater[rolloutGroup][imageName] ?? {
```
[the remark this answers is not in the document]
Comment (talkol) [reply src/model/state.ts:161]: it comes from the Ethereum events as one of the parameters emitted there, IMO it can't be undefined, but you tell me :)

File: src/model/state.ts [hunk 5]
```diff
@@ -148,14 +153,33 @@ export class State {
...
-    const currentVersion = this.snapshot.CurrentImageVersions[rolloutGroup][imageName];
-    // image version upgrades only go forward (we don't allow downgrade)
-    if (!currentVersion || Versioning.compare(imageVersion, currentVersion) > 0) {
-      this.snapshot.CurrentImageVersions[rolloutGroup][imageName] = imageVersion;
-    }
+  applyNewImageVersion(rolloutGroup: string, imageName: string, imageVersion: string) {
+    this.snapshot.CurrentImageVersions[rolloutGroup][imageName] = imageVersion;
+  }
+
+  applyNewImageVersionPollTime(time: number, rolloutGroup: string, imageName: string) {
```
[the remark this answers is not in the document]
Comment (talkol) [reply src/model/state.ts:160]: This is on purpose.. I wanted the parameter order to match `applyNewProtocolVersionChanged` and `applyNewSubscriptionChanged` that all receive the current time as first argument. `applyNewImageVersionPendingUpdate` is unique because the time there is not current time, but future time of the upgrade. It also had be in the end due to defaults..

File: src/model/state.ts [hunk 6]
```diff
@@ -148,14 +153,33 @@ export class State {
...
-  applyNewImageVersion(time: number, rolloutGroup: string, imageName: string, imageVersion: string) {
-    if (!Versioning.isValid(imageVersion)) return;
-    this.snapshot.CurrentImageVersionsUpdateTime[rolloutGroup][imageName] = time;
-    const currentVersion = this.snapshot.CurrentImageVersions[rolloutGroup][imageName];
-    // image version upgrades only go forward (we don't allow downgrade)
-    if (!currentVersion || Versioning.compare(imageVersion, currentVersion) > 0) {
-      this.snapshot.CurrentImageVersions[rolloutGroup][imageName] = imageVersion;
-    }
+  applyNewImageVersion(rolloutGroup: string, imageName: string, imageVersion: string) {
+    this.snapshot.CurrentImageVersions[rolloutGroup][imageName] = imageVersion;
```
[the remark this answers is not in the document]
Comment (talkol) [reply src/model/state.ts:157]: I clarified the reason for defaults in a comment

File: src/dockerhub/versioning.test.ts [hunk 7]
```diff
@@ -98,7 +105,32 @@ test('isCanary returns false on invalid versions', (t) => {
   }
 });
 
-test('copmpare sorts the latest version at the smallest index', (t) => {
+test('isHotfix returns false on invalid versions', (t) => {
```
[the remark this answers is not in the document]
Comment (talkol) [reply src/dockerhub/versioning.test.ts:108]: cool, added

File: src/model/state.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (ronnno) [reply src/model/state.ts:161]: i think you are right - its a minor point - the default initialized values `StateSnapshot` must match the hardcoded values in `fetchLatestVersion` for this assumption here to hold. it's a little scattered but i think in this version there is no bug...

**Comment by ronnno** [comment]
That was what I thought - my suggestion was to remove them from the snapshot because they have no functional value there and they appear to be stored there for functional reasons.

**Comment by talkol** [comment]
> That was what I thought - my suggestion was to remove them from the snapshot because they have no functional value there and they appear to be stored there for functional reasons.

hmm... the only sources of data for the endpoint renderers (`/node/management`, `/vchains/management`, `/status`) is the state and the config (which is static on launch). I'm not sure how elegantly to pass this data to the renderer if it's not in the state.

The way I look at state is all the current "persistent" state of the service. Eventually we will want to implement some persistence for this service so it will be able to restart without reconstructing the state from zero. The way to go IMO will be to write the state JSON to disk after every poll. When we do this, I will really want these fields there too, because I could reconstruct the timers of a pending upgrade based on this state information (meaning the service was restarted when there was a pending upgrade scheduled but not yet performed).

In other words, this information feels like the current state of the service.

The reason I also have the timers map in the `imagePoll` object is that it's a JS timer and it's not serializable so I couldn't put it inside the state.

File: src/api/render-status.ts [hunk 8]
```diff
@@ -50,7 +51,7 @@ function getErrorText(snapshot: StateSnapshot) {
     res.push(`Ethereum RefTime is too old (${refTimeAgo} sec ago).`);
   }
   for (const imageName of imageNamesToPollForNewVersions) {
-    const polledAgo = now - (snapshot.CurrentImageVersionsUpdateTime['main'][imageName] ?? 0);
+    const polledAgo = now - (snapshot.CurrentImageVersionsUpdater['main'][imageName]?.LastPollTime ?? 0);
```
[the remark this answers is not in the document]
Comment (talkol) [reply src/api/render-status.ts:54]: cool, added a separate check for `canary`

File: src/api/render-node.test.ts [hunk 9]
```diff
@@ -79,4 +81,35 @@ test.serial('[integration] getNodeManagement responds according to Ethereum and
   t.assert(res.orchestrator);
 
   scope.done();
+
+  // mock docker hub state with a few new versions
+  const scope2 = nockDockerHub(
```
[the remark this answers is not in the document]
Comment (talkol) [reply src/api/render-node.test.ts:86]: spelling mistake? lol :) no, the library that mocks HTTP is called "nock", so Amir called it `nockDockerHub` because it uses `nock` to mock it. 

https://www.npmjs.com/package/nock

File: src/dockerhub/image-poll.ts (hunk 1 above)
Comment (talkol) [reply src/dockerhub/image-poll.ts:64]: cool, added

**Comment by talkol** [comment]
The way I look at state is more of a "redux"/"mobx"-like state of the service, and not necessarily Ethereum state based on Ethereum events.

**Comment by ronnno** [comment]
i understand... just for the sake of argument - in this case I'd probably use a simple logic timer that wakes up every minute and checks the model for upcoming upgrades to implement. this way the service is simple and the state serves as source of truth.. but either way nice

**Comment by talkol** [comment]
Yeah, it's a good idea. I considered it instead of the `setTimeout` timer. Especially since `setTimeout` has a max limit of about 40 days. The problem was I wanted to poll this every second (current docker hub poll is every 3 minutes) so this meant creating a new task. We already have 2 main tasks (Eth block sync and docker hub poller), and adding another big task just for this was also ugly.. so I went for the timer option.. also another task means monitoring for the task, because it may die on exception..
